spring-integration-kafka/src/test/java/org/springframework/integration/kafka/inbound/MessageDrivenAdapterTests.java

```diff
@@ -281,6 +281,65 @@ protected boolean doSend(Message<?> message, long timeout) {
 		pf.reset();
 	}
 
+	@Test
+	void testInboundBatchRetryInListenerWithSharedRetryTemplate() {
+		RetryTemplate retryTemplate = new RetryTemplate();
+		SimpleRetryPolicy retryPolicy = new SimpleRetryPolicy();
+		retryPolicy.setMaxAttempts(2);
+		retryTemplate.setRetryPolicy(retryPolicy);
+
+		Map<String, Object> props1 = KafkaTestUtils.consumerProps(EMBEDDED_BROKERS, "test4", "true");
+		props1.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+		DefaultKafkaConsumerFactory<Integer, String> cf1 = new DefaultKafkaConsumerFactory<>(props1);
+		ContainerProperties containerProps1 = new ContainerProperties(topic4);
+		KafkaMessageListenerContainer<Integer, String> container1 =
+				new KafkaMessageListenerContainer<>(cf1, containerProps1);
+		KafkaMessageDrivenChannelAdapter<Integer, String> adapter1 = new KafkaMessageDrivenChannelAdapter<>(container1);
+		adapter1.setRetryTemplate(retryTemplate);
+		adapter1.afterPropertiesSet();
+
+		Map<String, Object> props = KafkaTestUtils.consumerProps(EMBEDDED_BROKERS, "test2", "true");
+		props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+		DefaultKafkaConsumerFactory<Integer, String> cf = new DefaultKafkaConsumerFactory<>(props);
+		ContainerProperties containerProps = new ContainerProperties(topic2);
+		containerProps.setIdleEventInterval(100L);
+		KafkaMessageListenerContainer<Integer, String> container =
+				new KafkaMessageListenerContainer<>(cf, containerProps);
+		KafkaMessageDrivenChannelAdapter<Integer, String> adapter = new KafkaMessageDrivenChannelAdapter<>(container,
+				ListenerMode.batch);
+		MessageChannel out = new DirectChannel() {
+
+			@Override
+			protected boolean doSend(Message<?> message, long timeout) {
+				return retryTemplate.execute(retryContext -> {
```

Comment:
Why do you use `RetryTemplate` in this piece of code?

Reply:
That usage would call `RetryListener.open()` method and reset  `ATTRIBUTES_HOLDER` finally.
 
This is what I tried to explain with https://github.com/spring-projects/spring-integration/pull/8771#issuecomment-1775707880, https://github.com/spring-projects/spring-integration/pull/8771#issuecomment-1773585667.

It's an edge case and a very specific and complex scenario. It will be a kind of burden to maintenance and the subject of this fix is changed. So, I removed the test.

Shall I add several lines to `MessageDrivenAdapterTests#testInboundRecordRetryRecover` as shown below?
```
	@Test
	void testInboundRecordRetryRecover() {
                .....
		adapter.setOutputChannel(out);
		RetryTemplate retryTemplate = spy(new RetryTemplate()); <-- this line
		.....
                assertThat(receivedMessageHistory.get().toString()).isEqualTo("myNullChannel");
		verify(retryTemplate, never()).registerListener(any()); <-- this line
		verify(retryTemplate, never()).registerListener(any(),anyInt()); <-- this line
                .....
	}
```
Or add a test similar to the one below?
```
	@Test
	void testShouldNotMutateRetryTemplate() {
		Map<String, Object> props = KafkaTestUtils.consumerProps(EMBEDDED_BROKERS, "test4", "true");
		props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
		DefaultKafkaConsumerFactory<Integer, String> cf = new DefaultKafkaConsumerFactory<>(props);
		ContainerProperties containerProps = new ContainerProperties(topic4);
		KafkaMessageListenerContainer<Integer, String> container =
				new KafkaMessageListenerContainer<>(cf, containerProps);
		KafkaMessageDrivenChannelAdapter<Integer, String> adapter = new KafkaMessageDrivenChannelAdapter<>(container);
		RetryTemplate retryTemplate = spy(new RetryTemplate());
		adapter.setRetryTemplate(retryTemplate);
		adapter.afterPropertiesSet();
		adapter.start();
		ContainerTestUtils.waitForAssignment(container, 1);
		
		verify(retryTemplate, never()).registerListener(any());
		verify(retryTemplate, never()).registerListener(any(),anyInt());

		adapter.stop();
	}
```